### DISCUSSION TITLE
How upload file from hook in Go?

### DISCUSSION THREAD

**Discussion by martinjanda** [pocketbase/pocketbase#1859] [Q&A] [answered]
I try use pocketbase as framework in Go.

Is there a easy way from the go (OnAfterBootstrap hook function) to upload a file as part of a record (via Dao record)?

Thank you so much

**Comment by martinjanda** [comment]
I'll try to add to the question. 

I have read files from another source as `[]byte`. I use local storage as file storage in pocketbase.

I can use the Web API, but that seems unnecessary. Is there a way to correctly create a file from Go code?

**Comment by ganigeorgiev** [comment, marked as answer]
This is in my todo to document it better in the near future, but in general there are 2 options depending how you want to create/update the record:

1. **Create/Update without data validations**
  This scenario uses the `Dao` directly, meaning that it is up to the developer to handle the "upload" to the filesystem and assign the filename to the record file field on save.

    ```go
    app.OnAfterBootstrap().Add(func(e *core.BootstrapEvent) error {
        yourFileBytes := []byte{...}

        collection, err := app.Dao().FindCollectionByNameOrId("yourCollection")
        if err != nil {
            return err
        }

        fs, err := app.NewFilesystem()
        if err != nil {
            return err
        }
        defer fs.Close()

        file, err := filesystem.NewFileFromBytes(yourFileBytes, "example_test.png")
        if err != nil {
            return err
        }

        // create new record OR fetch an existing one via https://pocketbase.io/docs/record-methods/#fetch-records
        record := models.NewRecord(collection)
        record.RefreshId()                         // generate a random id
        record.Set("yourFileFieldName", file.Name) // for multi-file fields is the same but as slice of strings

        fileKey := record.BaseFilesPath() + "/" + file.Name

        // upload the file
        if err := fs.UploadFile(file, fileKey); err != nil {
            return err
        }

        // persist the changes
        if err := app.Dao().SaveRecord(record); err != nil {
            // try to delete the previously uploaded file
            fs.Delete(fileKey)

            return err
        }

        return nil
    })
    ```
    _Note that you can wrap the above in a transaction with `app.Dao().RunInTransaction(...)` to avoid the manual file deletion on failure but we do it this way in order to minimize the transaction locking times, since the file upload could take some time for example if you use S3 as a storage (for a local files storage this is not much of a concern)._
    
2. **Create/Update with data validations**
   This approach uses the `forms.RecordUpsert` service. It performs field data validations, meaning that your file size and mime type will be checked against the file field constraints.

    ```go
    app.OnAfterBootstrap().Add(func(e *core.BootstrapEvent) error {
        yourFileBytes := []byte{...}

        collection, err := app.Dao().FindCollectionByNameOrId("yourCollection")
        if err != nil {
            return err
        }

        // create new record OR fetch an existing one via https://pocketbase.io/docs/record-methods/#fetch-records
        record := models.NewRecord(collection)

        form := forms.NewRecordUpsert(app, record)

        // set regular fields (if any)
        form.LoadData(map[string]any{
            "anyRegularField": 123,
        })

        file, err := filesystem.NewFileFromBytes(yourFileBytes, "example_test.png")
        if err != nil {
            return err
        }

        // add files to upload
        form.AddFiles("yourFileFieldName", file)

        // Submit handles validation, upload and record persistence
        return form.Submit()
    })
    ```


---

_I'm not sure what is your use case, but keep in mind that `OnAfterBootstrap` hook fires every time when the server starts, meaning that the above record creation will happen every time when you run `./pocketbase serve`. If you want to execute something only once, you can do it via [migrations](https://pocketbase.io/docs/migrations/) or create a console command via `app.RootCmd.AddCommand` and call it manually._

**Comment by martinjanda** [comment]
Thank you for fast answer.

in my case I want to create a gorutine in `OnAfterBootstrap` hook that will handle regular synchronization with an external data source.
